Starting specs 1 and 2 (#1)
pyviz/spec: spec_02.md

- Spec: 2
- Title: Methods on figure objects
- Version: 0.1
- Last-Modified: 2019-07-19
- Authors: [James Bednar](),
           [David Hoese](),
           [Jon Mease](jon.mease@gmail.com),
           [Julia Signell](jsignell@gmail.com),
- Status: Active
- Type: Standards
- Content-Type: text/markdown
- Created: 2019-07-19

## Abstract

> This spec describes a proposed shared API for accessing objects from various plotting libraries.

Reviewer: ```suggestion
This spec describes a proposed shared Python API for accessing objects from various plotting libraries.
```

> In this notebook we will use **figure** to refer to any object that has a visible representation but which can also be saved, exported, etc. This includes generic visualizations or animation that may or may not be a "plot" (axes, ticks, etc)?

Reviewer: ```suggestion
In this notebook we will use **figure** to refer to any object that has a visible representation but that can also be saved, exported, etc. This includes generic visualizations or animations that may or may not be considered a "plot" with axes, ticks, and so on.```

When detailing a method, this spec uses FIGURE to refer to any specific figure object and LIBRARY to refer to any specific visualization library.

## If a method doesn't make sense

> If a given operation doesn't make sense for that library, then it can satisfy the spec by simply having that method return a message to that effect ("JSON support not available in LIBRARY").

Reviewer: ```suggestion
If a given operation doesn't make sense for that library, then it can satisfy the spec by simply having that method return a message to that effect ("Unsupported: LIBRARY does not provide JSON output").
```

## Jupyter methods

> Every library that supports rendering in Jupyter should support the various IPython rich display methods, i.e. `repr_html`, `repr_png`, `_ipython_display_`.

Reviewer: ```suggestion
Every library that supports rendering in Jupyter should support at least one of the various IPython rich display methods, i.e. `_repr_html_`, `_repr_png_`, `_ipython_display_`.
```

## `.show()`
Every figure should have the ability to render itself.

### Proposal

>  1) The top-level figure class has a `.show()` method that can be called without arguments to display the figure as a side-effect.

Reviewer: ```suggestion
 1) The top-level figure class should have a `.show()` method that can be called without arguments to display the figure as a side-effect.
```

 2) The optional `renderer` kwarg can be used to override the current default renderer. e.g.:
    - `FIGURE.show(renderer='png')` to display the figure as a static png image/

>     - `FIGURE.show(renderer='browser')` to display the figure in a browser tab. This works in non-jupyter/IPython contexts.

Reviewer 2: What level of interactivity should be required from this `show` interface? And should there be a defined list of "preferred" rendering options that a library should support with defined names (ex. so one library doesn't use `jpg` while another uses `jpeg`).

Does `browser` mean interactive javascript based display? Or could it mean open a browser tab and display the PNG version of this figure? Does the browser rendering open a new tab if in a jupyter notebook environment or does it display it inline in the notebook cell's output?

Reviewer: I would argue for a minimal standard for `show` that can be met by all libraries, and thus would not assume any interactivity or JavaScript.

Author: should this maybe be `output` instead of `renderer` so that it can match `.save()`?

Author: should the `output` option just be `'html'` then and a browser tab is implicitly where html would be displayed?

Author: I think that makes sense.

## `.save()` method
Every figure should have the knowledge of how to export itself to a file on disk.

> #### What currently exists
>  - matplotlib has `.savefig` which saves the current figure, with options `fname` and `format` among others.
> - plotly has `write_*` methods which write the figure to a `file` (or writable object) and return `None`.
> - bokeh has top level `export_*` methods for each output format as well as `save` (only outputs html) which both take a FIGURE as the arg.

### Proposal
Figures should have `FIGURE.save()` method for exporting.

There are several kwargs that `save` should include:

>  1) `file`: should be a file object or optionally a path to a file. If a file path is used rather than a file object and that is not supported, the library should raise a sensible error. Can default to some user configured value.

Reviewer: ```suggestion
 1) `file`: should be a file object or optionally a path to a file. If the LIBRARY does not support file paths and a user provides one rather than a file object, the LIBRARY should raise a sensible error. Can default to some user configured value.
```

>  2) `output`: should be a file format that the figure can be exported to i.e. `'png'`, `'json'`, `'html'`. The library should set a default output and/or allow the user to configure the output. For instance matplotlib does:

Reviewer: ```suggestion
 2) `output`: should be a file format that the figure can be exported to i.e. `'png'`, `'jpg'`, `'json'`, `'html'`. The library should set a default output and/or allow the user to configure the output. For instance matplotlib does:
```

(Added jpg explicitly here to avoid having some libraries choose 'jpg' and others 'jpeg' for the same operation.)

- Spec: 2
- Title: Methods on figure objects
- Version: 0.1
- Last-Modified: 2019-07-19
- Authors: [James Bednar](),
           [David Hoese](),
           [Jon Mease](jon.mease@gmail.com),
           [Julia Signell](jsignell@gmail.com),
- Status: Active
- Type: Standards
- Content-Type: text/markdown
- Created: 2019-07-19

## Abstract
This spec describes a proposed shared Python API for accessing objects from various plotting libraries.

In this notebook we will use **figure** to refer to any object that has a visible representation but that can also be saved, exported, etc. This includes generic visualizations or animations that may or may not be considered a "plot" with axes, ticks, and so on.```

When detailing a method, this spec uses FIGURE to refer to any specific figure object and LIBRARY to refer to any specific visualization library.

## If a method doesn't make sense
If a given operation doesn't make sense for that library, then it can satisfy the spec by simply having that method return a message to that effect ("Unsupported: LIBRARY does not provide JSON output").

## Jupyter methods

> Every library that supports rendering in Jupyter should support at least one of the various IPython rich display methods, i.e. `_repr_html_`, `_repr_png_`, `_ipython_display_`.

Reviewer 3: I believe ``_repr_mimebundle_`` along with a separately-installed frontend extension is the recommended way to do this according to the Jupyter team.

Author: Do you propose just adding `_repr_mimebundle_` to the list, or enforcing that it should be implemented?

Reviewer 3: I think adding it to the list is fine, but folks from the Jupyter team may have stronger opinions.
